### PR TITLE
Better handling of module base_config_ext data

### PR DESCRIPTION
**Pull request by andyross**
This is a different take on a fix for #8639 than the workaround in #8575 on which it was originally coded.  It removes the requirement for components to have to muck with module internals[1] and places the complicated byte-offset math into just one place (in the module code that owns the struct format).  Really I think this is a lot cleaner.  If you guys want to merge the first version for clarity (given it's already "shipped"[2]), I can submit this as cleanup on top of that instead (which, again, is how this was written originally anyway).

With this, the big IPC4 workaround patch at the end of the AEC work is down to just one line, hope to have that cleaned up by tomorrow.  (see https://github.com/andyross/sof/commits/mtl-aec-tmp/ for the current status of my working tree).

[1] Basically, the first version: (1) requires every module to make a copy of the untyped "init_data" field of module_config, then (2) store that pointer **back** into a different field of module_config (because the first one points into the IPC command buffer and won't survive) where (3) it will be eventually used, **not** by the module code itself but by the ipc4/module layer that created it in the first place.  This avoids all that by having the module_adapter init code just parse and store the data at the start.

[2] In the MTL "stable" branch, where it's already been merged.  Can I just whine a bit about how frustrating it is to have to work across branches like this?  I didn't realize until I was about to submit that the code I was modifying wasn't actually upstream!

### CODE REVIEW

**Comment by andyross** [comment]
> I didn't realize until I was about to submit that the code I was modifying wasn't actually upstream!

As in, I wrote up a whole commit message carefully explaining and advocating my removal of an API that never merged, heh:

https://github.com/andyross/sof/commit/37f784f7af6369c77e3d09d9551dae620e2b7d2e

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (btian1) [inline src/audio/module_adapter/module_adapter_ipc4.c:68]: memcpy? since size are same.

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (btian1) [inline src/audio/module_adapter/module_adapter_ipc4.c:41]: cp is too simple, how about ipc_config_data?

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (btian1) [inline src/audio/module_adapter/module_adapter_ipc4.c:58]: suggest add some comment for this condition.

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (btian1) [inline src/audio/module_adapter/module_adapter_ipc4.c:66]: line 66 also need comments.

File: src/ipc/ipc4/helper.c
Comment (btian1) [inline src/ipc/ipc4/helper.c:490]: line 475 does not have one more space line, this can be removed as well for align.

File: src/ipc/ipc4/helper.c
Comment (kv2019i) [inline src/ipc/ipc4/helper.c:459]: Something missing in the patch, "struct processing_module" definition is not known and there are lot of fails in CI (defined in module/base.h). Now this is supposed to be private to module implementations, which is why I think @ranj063  opted to add a new property so it can be "set from outside".

**Review by kv2019i — COMMENTED** [review]
This does look cleaner. I'll let others chime in. Ext-cfg is optional, but given it's more common to have it, don't see why not to handle it in common code. There are some corner cases I'd like others to chime in.

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (kv2019i) [reply src/audio/module_adapter/module_adapter_ipc4.c:45]: It would seem this is ok. Most modules use "DECLARE_MODULE_ADAPTER()" to register the driver (which sets the type correctly) and library_manager calls "declare_dynamic_module_adapter(drv, SOF_COMP_MODULE_ADAPTER, ..)". So I don't see any use for IPC4 where the type would not be set to adapter. It still bugs me a bit a specific branch was added to cover this case and I can't find from git history any use for this, but maybe this was some variant of loadable modules. @jxstelter @ranj063 @RanderWang  @softwarecki @pjdobrowolski please do comment if I got this wrong.

File: src/ipc/ipc4/helper.c
Comment (ranj063) [reply src/ipc/ipc4/helper.c:459]: @andyross I do like the idea of parsing/saving the base_cfg_ext data in the module adapter instead of doing it in the individual modules. But I think you're going to have to add a new property to retrieve the extn data just like we do for the base_cfg data today.

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (andyross) [reply src/audio/module_adapter/module_adapter_ipc4.c:45]: That was my read, yeah.  The original had a check for that that then set the init_data unconditionally, which seems a little dangerous (what if it's not?  what's in the data and how do we prevent treating it like pin config?".  I grepped around and couldn't find any situations where it would happen, so I replaced it with an assert to be sure, figuring that CI would tell me if I missed something.

File: src/ipc/ipc4/helper.c
Comment (andyross) [reply src/ipc/ipc4/helper.c:459]: I see the failures, presumably that's just a header ordering issue?  Or if processing_module is an illegal header in some contexts, can you provide details?  It would be easy enough to wrap these behind an API.  That seems unlikely though, as the alternate code was already banging directly on the mod->priv.cfg.

@ranj063 you mean the COMP_ATTR_BASE_CONFIG_EXT from the version in your first PR?  That's actually deliberately omitted: the only spot that would use it is using the original/kernel-derived data in the new parsed fields.    The reason for making this a property would presumably be that components could then override it, but (and, OK, this is becoming a crusade of mine I guess) I think that's emphatically bad and should be avoided.  If topology lied and gave us an invalid value, the job of the firmware is to fail gracefully and get the integrator to fix the topology.

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (andyross) [reply src/audio/module_adapter/module_adapter_ipc4.c:68]: I actually agree as a point of style (the code computes the size once, validates against the input blob size, then memcpy's the amount of data it allocates).  Nonetheless there's a checkpatch warning about the use of bare memcpy(), so it seems like I can't win. :)

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (andyross) [reply src/audio/module_adapter/module_adapter_ipc4.c:52]: Well to be fair, this check is needed to prevent a buffer overrun during parsing.  It's true that there's also a typesafety issue in knowing "what kind of extra data do we have", but that's a protocol bug not really addressed by this PR :)

**Comment by andyross** [comment]
Cleaned up the backport from mtl-007-drop-stable (the only place I can test this) that rebases this on top of a reversion that should match main better.  Hopefully will fix the build issues.

File: src/ipc/ipc4/helper.c
Comment (ranj063) [reply src/ipc/ipc4/helper.c:464]: @andyross I think originally processing_module was intended to be private data accessible only for the module itself. But given that we are already accessing the base_cfg and base_cfg_ext are being accessed by the common code, I guess it makes sense to make it shared. By the same argument maybe we should remove the COMP_ATTR_BASE_CONFIG attribute as well.

File: src/ipc/ipc4/helper.c
Comment (ranj063) [inline src/ipc/ipc4/helper.c:506]: shouldn't this be changed to obs * 2? and similarly ibs * 2 below?

Comment (ranj063) [reply src/ipc/ipc4/helper.c:506]: but more simply isnt it enough to do MAX(ibs * 2, obs * 2)?

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (cujomalainey) [inline src/audio/module_adapter/module_adapter_ipc4.c:53]: little more verbose with names here please

File: src/audio/module_adapter/module_adapter.c
Comment (kv2019i) [reply src/audio/module_adapter/module_adapter.c:1339]: @andyross  ^^ We can do this in a follow-up as well, but in general we want to move the IPC specifc code in separate compilation units to keep the amount of ifdefs in generic code in check.

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (kv2019i) [reply src/audio/module_adapter/module_adapter_ipc4.c:66]: Hmm, I'm good with this even without comments. To follow this code, reader needs to have definition of the parsed binary blobs at hand.

**Review by kv2019i — APPROVED** [review]
CI looks good now. By directly accessing module internals is a change in convention, but we seem to have consensus this is ok in this case.

File: src/include/module/module/base.h
Comment (andyross) [reply src/include/module/module/base.h:33]: They're actually different values from different sources.  The fields in processing_module count the number of actual comp_buffer objects connected to the module.  The one here is a field set in an IPC message from the kernel based on topology input.   In the case where the base_cfg_ext is provided, they surely should be the same value.  But the extended config is actually not always present (some components have their own config structs here instead), so they might differ at runtime and we need to know how much memory was actually allocated behind the pointers here.

File: src/ipc/ipc4/helper.c
Comment (andyross) [reply src/ipc/ipc4/helper.c:464]: Agreed, I'll look at that as a followup cleanup.  The "kpb" component is the only spot I see in current code that's implementing the BASE_CONFIG attribute, and it's just echoing back the default base config anyway.

**Comment by andyross** [comment]
Grrr... sorry.  Came back to check CI and realized that I'd mis-squashed the naming fixup and put it in the wrong patch.  I've already torn my tree up for different work though so can't get back to fix it in the immediate term.  (I'm juggling like six PRs right now, heh).  Don't merge until I get the corrected version in

File: src/include/module/module/base.h
Comment (lyakh) [inline src/include/module/module/base.h:34]: nitpick: is "nb" an abbreviation for "number?" If so, it seems a bit non-standard to me, maybe just "n" or "num" or even "number_of..."

File: src/ipc/ipc4/helper.c
Comment (ranj063) [reply src/ipc/ipc4/helper.c:506]: @andyross not changing the buffer sizing according to the ibs/obs based on pin numbers would be incorrect isnt it? Let's assume we're binding modules with a non-zero pin ID's, what would sink_src_cfg.ibs or source_src_cfg.obs contain?

File: src/include/module/module/base.h
Comment (andyross) [reply src/include/module/module/base.h:34]: It's the pre-existing name for the field in struct ipc4_base_module_cfg_ext from which this is derived.  Personally I don't like it much either, but I'm trying to be a good citizen and not rock the boat too much.  :)

File: src/ipc/ipc4/helper.c
Comment (andyross) [reply src/ipc/ipc4/helper.c:506]: Oh.... gotcha.  I had missed that the calculation here was still done with the source_src_cfg value and not the recalculated-and-now-pin-aware values from above.   Interestingly, this patch actually started as a refactor of the code in mtl-007-drop-stable, and it actually has exactly the same bug: https://github.com/thesofproject/sof/blob/mtl-007-drop-stable/src/ipc/ipc4/helper.c#L494

Comment (ranj063) [reply src/ipc/ipc4/helper.c:506]: I dont think mtl-007 has the bug. in that patch, the buf_size is calculated from the ibs/obs set based on the pin ID i.e from the base config for pin ID 0 or from the base config extn if pin ID > 0

Comment (andyross) [reply src/ipc/ipc4/helper.c]: Should be fixed.  Please check.

**Review by ranj063 — APPROVED** [review]

**Review by jxstelter — APPROVED** [review]

**Comment by kv2019i** [comment]
@marcinszkudlinski ?

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (lgirdwood) [inline src/audio/module_adapter/module_adapter_ipc4.c:73]: This block could do with an inline comment describing what its doing.

File: src/audio/module_adapter/module_adapter_ipc4.c
Comment (lgirdwood) [reply src/audio/module_adapter/module_adapter_ipc4.c:45]: Is this really fatal ? I think its probably survivable where we return the error up the stack for this module/pipeline only and allow other pipelines/streams to continue.

**Review by lgirdwood — APPROVED** [review]
My changes can be fixed incrementally, I'd rather get this merged and tested more as priority.

**Review by lgirdwood — APPROVED** [review]
My changes can be fixed incrementally, I'd rather get this merged and tested more as priority.

**Review by marcinszkudlinski — APPROVED** [review]

**Comment by marc-hb** [comment]
Fix c3da4e3d623a91c for double-free regression was just merged in #8790

**Pull request merged**
